src/Build.UnitTests/BackEnd/TaskRouting_IntegrationTests.cs

```diff
@@ -0,0 +1,499 @@
+    
+    <Target Name=""TestTarget"">
+        <{taskName} />
+    </Target>
+</Project>";
+        }
+    }
+
+    /// <summary>
+    /// Helper utilities for testing task routing behavior.
+    /// Provides robust assertions that are less fragile than raw log string matching.
+    /// </summary>
+    internal static class TaskRoutingTestHelper
+    {
+        /// <summary>
+        /// Asserts that a task was launched in an external TaskHost process.
+        /// </summary>
+        /// <param name="logger">The build logger containing execution logs.</param>
+        /// <param name="taskName">The name of the task to verify.</param>
+        public static void AssertTaskUsedTaskHost(MockLogger logger, string taskName)
+        {
+            // Look for the distinctive "Launching task" message that indicates TaskHost usage
+            string launchingMessage = $"Launching task \"{taskName}\"";
+            logger.FullLog.ShouldContain(launchingMessage);
+            logger.FullLog.ShouldContain("external task host");
+        }
+
+        /// <summary>
+        /// Asserts that a task ran in-process (not in TaskHost).
+        /// </summary>
+        /// <param name="logger">The build logger containing execution logs.</param>
+        /// <param name="taskName">The name of the task to verify.</param>
+        public static void AssertTaskRanInProcess(MockLogger logger, string taskName)
+        {
+            // Verify the "Launching task" message does NOT appear for this task
+            string launchingMessage = $"Launching task \"{taskName}\"";
+            logger.FullLog.ShouldNotContain(launchingMessage);
+        }
+    }
+
+    #region Test Task Implementations
+
+    /// <summary>
+    /// NonEnlightened task without IMultiThreadableTask interface or MSBuildMultiThreadableTaskAttribute.
+    /// Should run in TaskHost in multi-threaded mode.
+    /// </summary>
+    public class NonEnlightenedTestTask : Task
+    {
+        public override bool Execute()
+        {
+            Log.LogMessage(MessageImportance.High, "NonEnlightenedTask executed");
+            return true;
+        }
+    }
+
+    /// <summary>
+    /// Task implementing IMultiThreadableTask interface.
+    /// Should run in-process in multi-threaded mode.
+    /// </summary>
+    public class InterfaceTestTask : Task, IMultiThreadableTask
+    {
+        public TaskEnvironment TaskEnvironment { get; set; }
+
+        public override bool Execute()
+        {
+            Log.LogMessage(MessageImportance.High, "TaskWithInterface executed");
+            return true;
+        }
+    }
+
+    #endregion
+}
+
+// Custom attribute definition in Microsoft.Build.Framework namespace to match what TaskRoutingDecision expects
+// TaskRoutingDecision looks for attributes with FullName = "Microsoft.Build.Framework.MSBuildMultiThreadableTaskAttribute"
+// Since the real attribute is internal in Framework, we define our own test version here
+namespace Microsoft.Build.Framework
```

Comment:
nit: you can move it to the end of the file, not keep it in the middle of `Microsoft.Build.Engine.UnitTests.BackEnd` namespace.